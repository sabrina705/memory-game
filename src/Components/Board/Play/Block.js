import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';

export class Block extends Component {
  static propTypes = {
    width: PropTypes.number.isRequired,
    index: PropTypes.number.isRequired,
    onClick: PropTypes.func.isRequired,
    selected: PropTypes.bool.isRequired,
    gem: PropTypes.bool.isRequired,
  };

  handleClick = () => {
    this.props.onClick(this.props.index);
  };

  render() {
    const { selected, width, gem } = this.props;
    const classes = classnames('block', { gem, selected });

    return (
      <div style={{ width }} onClick={this.handleClick} className={classes} />
    );
  }
}

export default Block;
