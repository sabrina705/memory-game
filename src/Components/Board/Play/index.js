import React, { Component, PropTypes } from 'react';
import { observer, inject } from 'mobx-react';
import classnames from 'classnames';
import { action } from 'mobx';
import { Link } from 'react-router';
import { findDOMNode } from 'react-dom';
import Block from './Block';
import leftBlocks from './lib/leftBlocks';

export class Play extends Component {
  static propTypes = {
    blocks: PropTypes.object.isRequired,
    length: PropTypes.number.isRequired
  };

  state = {
    gameOver: false,
    success: false
  };

  @action
  handleSelectBlock = (index) => {
    const { blocks, increaseLevel } = this.props;
    const block = blocks[index];
    block.selected = true;

    const left = leftBlocks(blocks).length;

    if (!left) {
      increaseLevel();
      setTimeout(() => {
        findDOMNode(this.refs['memorizeBtn']).click();
      }, 1000);
      this.setState({ success: true });
    } else if (block.gem !== true) {
      this.setState({ gameOver: true });
    }
  };

  renderBlock = (block, index) => {
    return <Block
      key={block.id}
      index={index}
      width={50}
      gem={block.gem}
      selected={block.selected}
      onClick={this.handleSelectBlock}
    />;
  };

  render() {
    const { blocks, length } = this.props;
    const { gameOver, success } = this.state;

    const l = length * 50;

    return (
      <div
        style={{ width: l, height: l, margin: '0 auto' }}
        className={classnames('Play', { gameOver })}
      >
        <div className="blocks">
          { blocks.map(this.renderBlock) }
        </div>

        {
          !gameOver && !success && <div>
            <center>Play</center>
          </div>
        }

        { gameOver && <div className='gameOver__overlay'>
          <div>
            Game Over <Link to="/memorize">Retry</Link>
          </div>
        </div> }

        { success && <div className='success__overlay'>
          <div>
            You are awesome <Link ref="memorizeBtn" to="/memorize">Next
            Level</Link>
          </div>
        </div> }
      </div>
    );
  }
}


export default inject(store => ({
  increaseLevel: store.board.increaseLevel.bind(store.board),
  length: store.board.currentLevel.length,
  blocks: store.board.blocks,
}))(observer(Play));
