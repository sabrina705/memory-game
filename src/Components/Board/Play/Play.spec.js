import React from 'react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import { Play } from '../Play';
import mobx from 'mobx';

const router = {
  push() {

  }
};

it('Play renders without crashing', () => {
  const div = document.createElement('div');
  const gems = mobx.observable([]);
  const blocks = mobx.observable([
    { id: 0, gem: false, selected: false },
    { id: 1, gem: true, selected: false },
  ]);

  ReactDOM.render((
    <Play
      router={router}
      gems={gems}
      blocks={blocks}
      length={3}
    />
  ), div);
});

it('Play/Block ', () => {
  const gems = mobx.observable([1,2,3]);
  const blocks = mobx.observable([
    { id: 0, gem: false, selected: false },
    { id: 1, gem: true, selected: false },
    { id: 2, gem: true, selected: false },
    { id: 3, gem: true, selected: false },
    { id: 4, gem: false, selected: false }
  ]);

  const component = renderer.create(
    <Play gems={gems} blocks={blocks} length={3} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree.children[0].children[0].props.onClick();
  tree = component.toJSON();

  expect(tree.children[0].children[0].props.className.includes('selected')).toBe(true);
});
