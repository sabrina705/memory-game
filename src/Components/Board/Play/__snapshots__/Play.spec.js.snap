exports[`test Play/Block  1`] = `
<div
  className="Play"
  style={
    Object {
      "height": 150,
      "width": 150,
    }
  }>
  <div
    className="blocks">
    <div
      className="block"
      onClick={[Function]}
      style={
        Object {
          "width": 50,
        }
      } />
    <div
      className="block gem"
      onClick={[Function]}
      style={
        Object {
          "width": 50,
        }
      } />
    <div
      className="block gem"
      onClick={[Function]}
      style={
        Object {
          "width": 50,
        }
      } />
    <div
      className="block gem"
      onClick={[Function]}
      style={
        Object {
          "width": 50,
        }
      } />
    <div
      className="block"
      onClick={[Function]}
      style={
        Object {
          "width": 50,
        }
      } />
  </div>
</div>
`;
